File: src/2015/02/day.02.ts
import fs from "fs";

import "@/utils/array";

type Present = [number, number, number];

function parseInput() {
  return fs
    .readFileSync("src/inputs/2015/02/day.02.input.txt")
    .toString()
    .split("\n")
    .filter((x) => x)
    .map(parsePresent);
}

function parsePresent(value: string) {
  return value.split("x").map(Number) as Present;
}

export function part1() {
  return getTotalPresentValue(calculateWrappingPaper);
}

export function part2() {
  return getTotalPresentValue(calculateRibbon);
}

function getTotalPresentValue(
  calculatePresentValue: (present: Present) => number
) {
  const presents = parseInput();

  return presents.map(calculatePresentValue).sum();
}

function calculateWrappingPaper([l, w, h]: Present) {
  return 2 * (l * w + w * h + h * l) + Math.min(l * w, w * h, h * l);
}

function calculateRibbon([l, w, h]: Present) {
  return l * w * h + 2 * Math.min(l + w, w + h, h + l);
}

File: src/2015/02/day.02.py
from timeit import default_timer as timer


def parse_input():
    with open("src/inputs/2015/02/day.02.input.txt", "r") as file:
        return [[int(x) for x in line[:-1].split("x")] for line in file]


def part1():
    dimensions = parse_input()
    return sum(map(required_wrapping_paper, dimensions))


def part2():
    dimensions = parse_input()
    return sum(map(required_ribbon, dimensions))


def required_wrapping_paper(lwh: list[int]):
    [l, w, h] = lwh
    return 2*l*w + 2*w*h + 2*h*l + min(l*w, w*h, h*l)


def required_ribbon(lwh: list[int]):
    [l, w, h] = lwh
    return min(2*(l+w), 2*(w+h), 2*(h+l)) + l*w*h


def timed(fn):
    start = timer()
    result = fn()

    return (timer() - start, result)


(p1_elapsed, p1_answer) = timed(part1)
print(f"Part 1: {p1_answer} ({p1_elapsed:.3f}s)")

(p2_elapsed, p2_answer) = timed(part2)
print(f"Part 2: {p2_answer} ({p2_elapsed:.3f}s)")
